File: deepeval/metrics/json_correctness/json_correctness.py
from typing import List, Optional, Union
import json
from pydantic import BaseModel, ValidationError

from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
    ConversationalTestCase,
)
from deepeval.metrics import BaseMetric
from deepeval.metrics.utils import (
    construct_verbose_logs,
    check_llm_test_case_params,
    initialize_model,
    trimAndLoadJson,
)
from deepeval.models import DeepEvalBaseLLM
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.metrics.json_correctness.template import JsonCorrectnessTemplate
from deepeval.metrics.json_correctness.schema import Reason
from deepeval.utils import get_or_create_event_loop

DEFAULT_CORRERCT_REASON = "The generated Json matches and is syntactically correct to the expected schema."


class JsonCorrectnessMetric(BaseMetric):
    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
    ]

    def __init__(
        self,
        expected_schema: BaseModel,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        threshold: float = 0.5,
        async_mode: bool = True,
        include_reason: bool = True,
        strict_mode: bool = True,
        verbose_mode: bool = False,
    ):
        self.threshold = 1 if strict_mode else threshold
        self.model, self.using_native_model = initialize_model(model)
        self.include_reason = include_reason
        self.strict_mode = strict_mode
        self.async_mode = async_mode
        self.verbose_mode = verbose_mode
        self.expected_schema = expected_schema

    def measure(
        self,
        test_case: Union[LLMTestCase, ConversationalTestCase],
        _show_indicator: bool = True,
    ) -> float:
        if isinstance(test_case, ConversationalTestCase):
            test_case = test_case.turns[-1]
        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(self, _show_indicator=_show_indicator):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(test_case, _show_indicator=False)
                )
            else:
                valid_json = True
                try:
                    self.expected_schema.model_validate_json(
                        test_case.actual_output
                    )
                except ValidationError as e:
                    valid_json = False

                self.score = 1 if valid_json else 0
                self.reason = self.generate_reason(test_case.actual_output)
                self.success = self.score >= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"LLM outputed Json:\n{test_case.actual_output}",
                        # f"Expected Json Schema:\n{json.dumps(self.expected_schema.model_json_schema(), indent=4)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )

                return self.score

    async def a_measure(
        self,
        test_case: Union[LLMTestCase, ConversationalTestCase],
        _show_indicator: bool = True,
    ) -> float:
        if isinstance(test_case, ConversationalTestCase):
            test_case = test_case.turns[-1]
        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, async_mode=True, _show_indicator=_show_indicator
        ):
            valid_json = True
            try:
                self.expected_schema.model_validate_json(
                    test_case.actual_output
                )
            except ValidationError as e:
                valid_json = False

            self.score = 1 if valid_json else 0
            self.reason = await self.a_generate_reason(test_case.actual_output)
            self.success = self.score >= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"LLM outputed Json:\n{test_case.actual_output}",
                    # f"Expected Json Schema:\n{json.dumps(self.expected_schema.model_json_schema(), indent=4)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )

            return self.score

    async def a_generate_reason(self, actual_output: str) -> str:
        if self.include_reason is False:
            return None

        is_valid_json = self.score == 1
        if is_valid_json:
            return DEFAULT_CORRERCT_REASON

        prompt: dict = JsonCorrectnessTemplate.generate_reason(
            actual_output=actual_output,
            expected_schema=json.dumps(
                self.expected_schema.model_json_schema(), indent=4
            ),
            is_valid_json=is_valid_json,
        )

        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Reason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: Reason = await self.model.a_generate(prompt, schema=Reason)
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def generate_reason(self, actual_output: str) -> str:
        if self.include_reason is False:
            return None

        is_valid_json = self.score == 1
        if is_valid_json:
            return DEFAULT_CORRERCT_REASON

        prompt: dict = JsonCorrectnessTemplate.generate_reason(
            actual_output=actual_output,
            expected_schema=json.dumps(
                self.expected_schema.model_json_schema(), indent=4
            ),
            is_valid_json=is_valid_json,
        )

        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Reason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: Reason = self.model.generate(prompt, schema=Reason)
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score >= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Json Correctness"

File: typescript/src/metrics/json-correctness/json-correctness.ts
import type { ZodType } from "zod";
import { BaseMetric, resolveThreshold } from "@/metrics/base-metrics";
import { LLMTestCase, SingleTurnParams } from "@/test-case";
import { DeepEvalBaseLLM } from "@/models";
import type { EvalModeName } from "@/config/eval-mode";
import { toJsonSchema } from "@/models/utils";
import {
  initializeMetricModels,
  generateWithSchema,
  checkSingleTurnParams,
  constructVerboseLogs,
} from "@/metrics/utils";
import { JsonCorrectnessScoreReasonSchema } from "@/metrics/json-correctness/schema";
import { type MetricTemplateOverride } from "@/templates/override";

const TEMPLATE_CLASS = "JsonCorrectnessMetric";

export type JsonCorrectnessTemplateOverride =
  MetricTemplateOverride<"JsonCorrectnessMetric">;
const DEFAULT_CORRECT_REASON =
  "The generated Json matches and is syntactically correct to the expected schema.";

export interface JsonCorrectnessMetricOptions {
  /** The schema the output must conform to (zod ⇄ Python's pydantic model). Required. */
  expectedSchema: ZodType;
  threshold?: number | null;
  flaky?: boolean;
  model?: DeepEvalBaseLLM | string;
  /** Who decides; defaults to `DEEPEVAL_EVAL_MODE`, then `llm`. */
  evalMode?: EvalModeName;
  includeReason?: boolean;
  /** Defaults to true (matches Python): requires a perfectly valid JSON. */
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  evaluationTemplate?: JsonCorrectnessTemplateOverride;
}

/**
 * Json Correctness — does the `actualOutput` parse as JSON and conform to
 * `expectedSchema`? Deterministic check (score 1/0); an LLM only explains
 * failures. **Higher is better** (`success = score >= threshold`).
 */
export class JsonCorrectnessMetric extends BaseMetric {
  private readonly expectedSchema: ZodType;
  private validationError?: string;

  constructor(options: JsonCorrectnessMetricOptions) {
    const strictMode = options.strictMode ?? true;
    super(strictMode ? 1 : resolveThreshold(options.threshold, 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
      flaky: options.flaky,
      evaluationTemplate: options.evaluationTemplate,
    });
    this.templateClass = TEMPLATE_CLASS;
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.ACTUAL_OUTPUT,
    ];
    this.expectedSchema = options.expectedSchema;
    initializeMetricModels(this, {
      model: options.model,
      evalMode: options.evalMode,
      systemOne: false,
    });
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      checkSingleTurnParams(testCase, this.requiredParams, this);
      this.evaluationCost = this.usingNativeModel ? 0 : undefined;

      const validJson = this.isValidJson(testCase.actualOutput);
      this.score = validJson ? 1 : 0;
      this.reason = await this.generateReason(testCase.actualOutput);
      this.success = this.isSuccessful();

      this.verboseLogs = constructVerboseLogs(this, [
        `LLM outputed Json:\n${testCase.actualOutput}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  private isValidJson(actualOutput: string): boolean {
    this.validationError = undefined;
    try {
      this.expectedSchema.parse(JSON.parse(actualOutput));
      return true;
    } catch (e) {
      this.validationError = e instanceof Error ? e.message : String(e);
      return false;
    }
  }

  private async generateReason(
    actualOutput: string,
  ): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    if (this.score === 1) return DEFAULT_CORRECT_REASON;
    if (!this.model) return this.validationError;

    const prompt = this.getPrompt("generate_reason", {
      actual_output: actualOutput,
      expected_schema: JSON.stringify(
        toJsonSchema(this.expectedSchema),
        null,
        4,
      ),
      is_valid_json: false,
    });
    const { reason } = await generateWithSchema(
      this,
      prompt,
      JsonCorrectnessScoreReasonSchema,
    );
    return reason;
  }

  get name(): string {
    return "Json Correctness";
  }
}
